framework.cpp

```diff
@@ -0,0 +1,403 @@
+#ifndef _METRIC_FRAMEWORK_CPP
+#define _METRIC_FRAMEWORK_CPP
+
+#include <cmath>
+#include <vector>
+//#include <map>
+
+#include "3rdparty/blaze/Blaze.h"
+
+
+namespace metric {
+
+
+
+
+// --------
+// common type trait metafunctions
+
+
+
+template <typename>
+struct determine_container_type  // checks whether container is STL container (1) or Blaze vector (2)
+{
+    constexpr static int code = 0;
+};
+
+template <template <typename, typename> class Container, typename ValueType, typename Allocator>
+struct determine_container_type<Container<ValueType, Allocator>>
+{
+    constexpr static int code = 1; // STL
+};
+
+template <template <typename, bool> class Container, typename ValueType, bool F>
+struct determine_container_type<Container<ValueType, F>>
+{
+    constexpr static int code = 2; // blaze::DynamicVector
+};
+
+template <template <typename, int, int, int, int, int> class Container, typename ValueType, int _Rows, int _Cols, int _Options, int _MaxRows, int _MaxCols>
+struct determine_container_type<Container<ValueType, _Rows, _Cols, _Options, _MaxRows, _MaxCols>>
+{
+    constexpr static int code = 3; // Eigen::Array
+};
+
+
+template<typename C, int = determine_container_type<C>::code>
+struct determine_val_t  // determines type of element both for STL containers and Blaze vectors
+{
+    using type = void;
+};
+
+template<typename C>
+struct determine_val_t<C, 1>
+{
+    using type = typename C::value_type;
+};
+
+template<typename C>
+struct determine_val_t<C, 2>
+{
+    using type = typename C::ElementType;
+};
+
+template<typename C>
+struct determine_val_t<C, 3>
+{
+    using type = typename C::Scalar;
+};
+
+
+
+// --------
+// base classes
+
+
+template <typename rec_t>
+class MetricBase { // all metrics to be derived from
+
+public:
+
+    using Rec_T = rec_t;
+    using Val_T = determine_val_t<Rec_T>;
+
+    MetricBase<rec_t>() {}
+
+    double operator () (const rec_t & a, const rec_t & b) { // to be overridden in derived metrics
+        return 0;
+    }
+
+private:
+
+};
+
+
+
+
+template <typename Metric = MetricBase<double>>
```

Comment:
it says double, but how we will achieve this, i.e. in your example `RecType` is a type of container

Reply:
we can set defaults
metric = euclidean
record = vector
values = double
container = vector

Reply:
this is my bug left from container container code. Here default shoud be a container, e g vector<double>

Reply:
also without template template approach (where i can mass record container type independently on value type), i'm not sure default type vector<double> is needed at all

Reply:
@maxfilippov @mkh-welsch Ideally it should be possible to specify everywhere any vector-like container, so we can easily project numpy arrays

Reply:
@jura-gresko you mean we always provide an overload (or converter) that returns data in such a container for Python purpose? ok, lets add it to the standard. Can you please provide a simple example to follow?

Reply:
@maxfilippov any vector-like specialization will work, it is not required to return data in the same format in case of python wrapper, I can make opaque wrapper to just represent returned value in python, such object can be passed to any C++ function without copying its value.
An example of vector-like adapter for Numpy Array
```cpp
template<typename T>
class NumpyToVectorAdapter: public pybind11::array_t<T> {
public:
    using pybind11::array_t<T>::array_t;

    NumpyToVectorAdapter(pybind11::array_t<T> obj)
        : pybind11::array_t<T>(obj)
    {
    }

    bool empty() const {return this->size() == 0;}

    T* begin() const {
        return pybind11::detail::array_begin<T>(this->request());
    }

    T* end() const {
        return pybind11::detail::array_end<T>(this->request());
    }

    const T& operator[](size_t index) const {
        auto r = pybind11::array_t<T>::template unchecked<1>();
        return r[index];
    }
};
```